cuda: Add 12.8 with Blackwell support

pauleonix opened the pull request:
- Adds `100`, `100a`, `101`, `101a`, `120` and `120a` compute capabilities.
- Now supports gcc@14 and clang@19 host compilers.
- I added myself to the maintainers comment in `lib/spack/spack/build_systems/cuda.py` (I only noticed the comment today and I was already a maintainer of the cuda package and have been modifying the build system code with every release).

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

pauleonix committed:
cuda: Add 12.8 with Blackwell support

ax3l reviewed (dismissed):
Thank you!

pauleonix commented:
Regarding the CI build failures of PETSc, VTK-m and cuSZ:
- I informed cuSZ of their incompatibility due to missing includes in https://github.com/szcompressor/cuSZ/issues/84
- VTK-m is a linker error due to undefined references to VTK-m internal functions (no CUDA symbols).
- PETSc is a configuration failure. The configuration log contains a ton of errors, most due to missing files (not CUDA related).

balay commented:
https://gitlab.spack.io/spack/spack/-/jobs/14776314

```
Executing: /home/software/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeh/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/cuda-12.8.0-rcag6lsmyn2irfllc77ehir3lspr24px/bin/nvcc -c -o /tmp/petsc-y5qpobcg/config.setCompilers/conftest.o -I/tmp/petsc-y5qpobcg/config.setCompilers   /tmp/petsc-y5qpobcg/config.setCompilers/conftest.cu 
stdout: nvcc warning : Support for offline compilation for architectures prior to '<compute/sm/lto>_75' will be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
Possible ERROR while running compiler:exit code 0
```

Looks like this new version of nvcc is always printing this warning to stdout - this is causing grief with petsc configure

cc: @jczhang07

jczhang07 commented:
@balay what is the content of `conftest.cu`?

balay commented:
```
balay@petsc-gpu-01:/scratch/balay/petsc$ touch a.cu
balay@petsc-gpu-01:/scratch/balay/petsc$ /scratch/balay/spack/opt/spack/linux-ubuntu22.04-x86_64/gcc-11.4.0/cuda-12.8.0-ockidy37bxdqb7c3tclf3pkkrfmsarzu/bin/nvcc -c a.cu 
nvcc warning : Support for offline compilation for architectures prior to '<compute/sm/lto>_75' will be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
balay@petsc-gpu-01:/scratch/balay/petsc$ 
```
This version of compiler is always verbose on stdout :(

balay commented:
One workaround:
```
balay@petsc-gpu-01:/scratch/balay/petsc$ ./configure --with-cuda-dir=/scratch/balay/spack/opt/spack/linux-ubuntu22.04-x86_64/gcc-11.4.0/cuda-12.8.0-ockidy37bxdqb7c3tclf3pkkrfmsarzu CUDAPPFLAGS=-Wno-deprecated-gpu-targets
```

balay commented:
patch to configure to workaround:

https://gitlab.com/petsc/petsc/-/merge_requests/8095

balay committed:
petsc: add workaround to verbose nvcc

balay@petsc-gpu-01:/scratch/balay/petsc$ touch a.cu
balay@petsc-gpu-01:/scratch/balay/petsc$ /scratch/balay/spack/opt/spack/linux-ubuntu22.04-x86_64/gcc-11.4.0/cuda-12.8.0-ockidy37bxdqb7c3tclf3pkkrfmsarzu/bin/nvcc -c a.cu
nvcc warning : Support for offline compilation for architectures prior to '<compute/sm/lto>_75' will be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
balay@petsc-gpu-01:/scratch/balay/petsc$

balay committed:
Revert "petsc: add workaround to verbose nvcc"

This reverts commit ef2faa204669eb763c5d74cbe9d2ac448529d452.

balay committed:
petsc: add workaround to verbose nvcc (limit only to version 12.8.0 - assuming 12.8.1 won't be similarly verbose)

balay@petsc-gpu-01:/scratch/balay/petsc$ touch a.cu
balay@petsc-gpu-01:/scratch/balay/petsc$ /scratch/balay/spack/opt/spack/linux-ubuntu22.04-x86_64/gcc-11.4.0/cuda-12.8.0-ockidy37bxdqb7c3tclf3pkkrfmsarzu/bin/nvcc -c a.cu
nvcc warning : Support for offline compilation for architectures prior to '<compute/sm/lto>_75' will be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
balay@petsc-gpu-01:/scratch/balay/petsc$

balay commented:
Hm - since the above patch won't apply to older releases - perhaps best to use CUDAPPFLAGS in spack.

So pushed this update.

pauleonix commented:
Ugh cuSZ versioning is a mess. `cusz@0.6` (the default) in spack references a commit that is newer than the `v0.6` tag in their repo. In fact that commit is tagged `v0.9-rc1` in their repo...

haampie commented:
6 kinda looks like a 9, so probably a typo. Spack supports rc versions, change it to `version("0.9rc1", commit=...)`

haampie commented:
If you care then you could duplicate the version with same commit but add deprecated=True to the old one.

pauleonix committed:
cusz: Fix erroneous version number

pauleonix committed:
cusz: Add newer versions and patches for cuda@12.8

pauleonix commented:
libpressio only built with cusz@0.9 (release candidates), so I changed its dependency accordingly when fixing the versioning.

robertu94 commented:
> Ugh cuSZ versioning is a mess. `cusz@0.6` (the default) in spack references a commit that is newer than the `v0.6` tag in their repo. In fact that commit is tagged `v0.9-rc1` in their repo. There is no final `v0.9` tag or release. But there are newer release candidates for `v0.9` and newer release tags up to `v0.14`...
> 
> @robertu94 you seem to have added that commit SHA as `cusz@0.6` in #44076. I guess by mistake? Not sure where to go from here as correcting the commit that `cusz@0.6` refers to in spack could break stuff depending on it, I guess.


Sorry about that, the student who manages the tags in the repo changes them periodically by force pushing;  I've asked him to stop doing that going forward.  The commits added are the ones known to work and should not be changed.

robertu94 commented:
> @robertu94 libpressio only built with cusz@0.9 release candidates (i.e. neither the actual cusz@0.6 nor cusz@0.14 work), so I changed its dependency accordingly when fixing the versioning.
> 
> I completely removed cusz@0.6 and added a comment to avoid confusion.

Were actually in the midst of staging a bunch of updates to compression packages, but I am waiting for a few packages (including cuSZ to stabilize) before we push them upstream.  I'm okay removing older versions in the meantime.

https://github.com/robertu94/spack_packages/blob/master/packages/cusz/package.py

pauleonix commented:
> > Ugh cuSZ versioning is a mess. `cusz@0.6` (the default) in spack references a commit that is newer than the `v0.6` tag in their repo. In fact that commit is tagged `v0.9-rc1` in their repo. There is no final `v0.9` tag or release. But there are newer release candidates for `v0.9` and newer release tags up to `v0.14`...
> > @robertu94 you seem to have added that commit SHA as `cusz@0.6` in #44076. I guess by mistake? Not sure where to go from here as correcting the commit that `cusz@0.6` refers to in spack could break stuff depending on it, I guess.
> 
> Sorry about that, the postdoc who manages the tags in that repo changes them periodically by force pushing; I've asked him to stop doing that going forward. The commits added are the ones known to work and should not be changed.

Sounds like he should use branches for that like the spack release branches that allow for patches to be added.

robertu94 commented:
> Sounds like he should use branches for that like the spack release branches that allow for patches to be added.

or just don't force push anything including both tags and branches... It's been said.

pauleonix commented:
Ok, the vtk issue seemingly went away by itself... But now py-jaxlib fails to build. @adamjstewart any idea what is going on? I saw your issue in jax which might be related?

adamjstewart commented:
It could be that JAX fails to build on all pipelines since https://github.com/spack/spack/pull/46346 was merged, not sure. If you notice this build failure on other PRs, let's revert that PR.

pauleonix commented:
@adamjstewart Locally I can reproduce the failure with `py-jaxlib^cuda@12.8.0` but `py-jaxlib^cuda@12.6.3` does build, so I don't think it will fail in other pipelines. Should I just add a conflict to `py-jaxlib`?

adamjstewart commented:
Oh good, that means I didn't break anything. Yes, let's just add a conflict. If you're feeling ambitious, you could open an issue with the JAX developers to see if/when CUDA 12.8 support was added, but I'm guessing it works in newer releases.

pauleonix commented:
@adamjstewart I would just make the conflict specific to the currently available versions in spack, so whenever newer versions are added, there should be another shot at compiling it with cuda@12.8 or newer.

pauleonix committed:
py-jaxlib: Add conflict with cuda@12.8:

pauleonix committed:
Merge branch 'develop' into cuda12.8

pauleonix commented:
@spackbot fix style

pauleonix committed:
[@spackbot] updating style on behalf of pauleonix

adamjstewart reviewed (approved):
CUDA and JAX packages look good, didn't review the others.
